subprojects/docs/src/docs/userguide/dependencyManagement.adoc

```diff
@@ -198,102 +198,151 @@ Let's say you wanted to declare a dependency on the link:https://tomcat.apache.o
+Gradle resolves any dependency version conflicts by selecting the latest version found in the dependency graph. Some projects might need to divert from the default behavior and enforce an earlier version of a dependency e.g. if the source code of the project depends on an older API of a dependency than some of the external libraries.
 
-* The dependency is undesired due to licensing reasons.
-* The dependency is not available in any remote repositories.
-* The dependency is not needed for runtime.
-* The dependency has a version that conflicts with a desired version. For that use case please refer to <<sub:version_conflicts>> and the documentation on api:org.gradle.api.artifacts.ResolutionStrategy[] for a potentially better solution to the problem.
+[NOTE]
+====
+Enforcing a version of a dependency requires a conscious decision. Changing the version of a transitive dependency might lead to runtime errors if external libraries do not properly function without them. Consider upgrading your source code to use a newer version of the library as an alternative approach.
+====
+
+Let's say a project uses the link:https://hc.apache.org/httpcomponents-client-ga/[HttpClient library] for performing HTTP calls. HttpClient pulls in link:https://commons.apache.org/proper/commons-codec/[Commons Codec] as transitive dependency with version 1.10. However, the production source code of the project requires an API from Commons Codec 1.9 which is not available in 1.10 anymore. A dependency version can be enforced by declaring it in the build script and setting api:org.gradle.api.artifacts.ExternalDependency#setForce(boolean)[] to `true`.
+
+++++
+<sample id="force-per-dependency" dir="userguide/dependencies/forcingDependencyVersion" title="Enforcing a dependency version">
+    <sourcefile file="build.gradle" snippet="force-per-dependency"/>
+</sample>
+++++
 
-Basically, in most of the cases excluding the transitive dependency should be done per configuration. This way the dependency declaration is more explicit. It is also more accurate because a per-dependency exclude rule does not guarantee the given transitive dependency does not show up in the configuration. For example, some other dependency, which does not have any exclude rules, might pull in that unwanted transitive dependency.
+If the project requires a specific version of a dependency on a configuration-level then it can be achieved by calling the method api:org.gradle.api.artifacts.ResolutionStrategy#force(java.lang.Object...)[]. Forced dependencies on the configuration-level only work if the configuration is resolvable; that is api:org.gradle.api.artifacts.Configuration#isCanBeResolved()[] returns `true`.
```

Comment:
"if the configuration is resolvable" — I don't know what this means. In what user-facing scenario would a configuration not be "resolvable"?

Reply:
This refers to the configurations defined and described in the [Java Libray plugin](https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_configurations_graph). I'll be honest I am not sure what the exact design decision was here. The documentation doesn't really explain it. The same is true for "consumable".

@jjohannes @melix Could you please clarify?

Reply:
Thanks for bringing this up @eriwen. I also stumbled over this in my review. But I did not have a good idea how to improve it then. 

A general problem I think is that too much "implementation details" about _configurations_ bleed into the API atm. This, again, has historical reasons. _Configuration_ became such a central concept in the implementation, that it is responsible for "everything" in dependency resoultion.
(The long term fix for this will be to turn the whole configuration concept more and more into an internal detail and providing new user facing APIs that talk about _variants_ instead of _configurations_. I think we will eventually get there in a year or so.)

For now, we have two kinds of configurations
1. Configurations to declare dependencies (these are now usually **not** _resolvable_)
2. Configurations to consume (these are _resolvable_)

This is, because configurations can be combined (via _extends_). So, for example, we have a couple of configs to declare dependencies that end up in the compile classpath. But we have only one _compileClasspath_ configuration that combines them all.
Because in the end, a consumer only cares about the complete classpath. And not the details, where exactly each dependency was declared.

So what can we do here? Maybe add an explanation at the place where configurations are introduced. Paraphrasing what I wrote above. Something along the lines of:
_There are two kind of configurations. One to declare dependencies and one that is exposed to tasks in the build to consume dependencies. For example, dependencies you require to compile a java project are defined using the `implementation` configuration. The compile task then uses the `compileClasspath` configuration which combines the dependencies defined in `implementation` with all other dependencies needed to compile the project (e.g. transitive dependencies)_

I have mixed feeling about investing too much effort to explaining this better. Because eventually I think the API should be improved by leveraging variant aware dependency management. But that will not happen soon. So it's probably worth documenting this better now.

Reply:
Thank you for the explanation @jjohannes. So concretely: `compile` is not "resolvable" but .. wait now can you clarify exactly what you mean by "Configurations to consume"? A concrete example would be the very best way to explain here, I think.

Folks are going to have to deal with this for many years until they can upgrade to 6.0 or later. We _must_ document this. Clearly. More documentation is the price we pay for developing complex concepts.

Reply:
Thanks for the details, @jjohannes. I am going to remove the comments here for now. The meaning of the terms should rather be documented in a new section that explains it more generally. The implications also need to be explained in the docs on the `java` and `java-library` plugin. I will work on this soon.